### PR TITLE
Adding Flame test use case methods

### PR DESCRIPTION
**Pull request by erickzanardo**
# Description

This PR adds two new methods to `flame_test`, they add a (optional) way of testing flame in an easier way. Inspired by [bloc_test](https://pub.dev/packages/bloc_test).

For now they are quite simple and just abstract few details of code needed to test flame games, but we can elaborate more in the future.

The `flameWidgetTest` specifically enables images to be loaded inside a flutter widget test environment, which fixes: #790.

As a bonus, this PR adds a few tests to Flame's own code base to demonstrate the new added methods.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors and are passing (See [Contributor Guide]).
- [x] My PR does not decrease the code coverage, or I have __a very special case__ and explained on the PR description why this PR decreases the coverage.
- [ ] I updated/added relevant documentation (doc comments with `///`) and updated/added examples in `doc/examples`.
- [x] I have formatted my code with `./scripts/format.sh` and the Flame analyzer (`./scripts/analyze.sh`) does not report any problems.
- [x] I read and followed the [Flame Style Guide].
- [x] I have added a description of the change under `[next]` in `CHANGELOG.md`.
- [x] I removed the `Draft` status, by clicking on the `Ready for review` button in this PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flame users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in `CHANGELOG.md`).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flame-engine/flame/issues
[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Flame Style Guide]: https://github.com/flame-engine/flame/blob/main/STYLEGUIDE.md

### CODE REVIEW

File: packages/flame_test/lib/src/flame_test.dart [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+void flameWidgetTest<T extends Game>(
+  String description, {
+  required GameCreateFunction<T> createGame,
+  GameWidgetCreateFunction<T>? createGameWidget,
+  WidgetVerifyFunction<T>? verify,
+}) {
+  testWidgets(description, (tester) async {
+    final game = createGame();
+
+    await tester.runAsync(() async {
```
Comment (erickzanardo) [inline packages/flame_test/lib/src/flame_test.dart:70]: this method `tester.runAsync` is what allows the images to be loaded inside the Flutter Widget testing environment.

File: packages/flame_test/lib/src/flame_test.dart [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+typedef GameWidgetCreateFunction<T extends Game> = GameWidget<T> Function(
+    T game);
+typedef WidgetVerifyFunction<T extends Game> = Future<void> Function(
+  T,
+  WidgetTester,
+);
+
+/// Creates a [Game] specific test case with given [description]
+/// which runs inside the flutter inside the Flutter test enviroment.
```
Comment (erickzanardo) [inline packages/flame_test/lib/src/flame_test.dart:51]:
```suggestion
/// which runs inside the Flutter test environment.
```

**Review by spydon — APPROVED** [review]
Nice!!

File: packages/flame_test/lib/src/flame_test.dart [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+/// Creates a [Game] specific test case with given [description].
+///
+/// Use [createGame] to create your game instance.
+///
+/// Use [verify] closure to make verifications/assertions.
+@isTest
+void flameTest<T extends Game>(
+  String description, {
+  required GameCreateFunction<T> createGame,
+  required VerifyFunction<T>? verify,
```
Comment (st-pasha) [inline packages/flame_test/lib/src/flame_test.dart:25]: If `verify` is nullable, then why is it required?

Comment (erickzanardo) [inline packages/flame_test/lib/src/flame_test.dart:25]:
```suggestion
  VerifyFunction<T>? verify,
```

Comment (erickzanardo) [reply packages/flame_test/lib/src/flame_test.dart]: I forgot to removed the required, removed now

File: packages/flame_test/example/lib/game.dart [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+  Widget build(BuildContext context) {
+    return GameWidget(game: MyGame());
+  }
+}
+
+class MyGame extends FlameGame {
+  @override
+  Future<void> onLoad() async {
+    final citySprite = await loadSprite('city.png');
+    await add(SpriteComponent(sprite: citySprite, size: Vector2(200, 200)));
```
Comment (luanpotter) [inline packages/flame_test/example/lib/game.dart:16]: nit: Vector2.all(200)

File: packages/flame_test/CHANGELOG.md [hunk 5]
```diff
@@ -1,5 +1,8 @@
 # CHANGELOG
 
+## [next]
+ - Adding `flameTest` and `flameWidgetTest`
```
Comment (luanpotter) [inline packages/flame_test/CHANGELOG.md:4]: nit: `Add` (to be consistent with the other entry below)

**Review by luanpotter — APPROVED** [review]
LGTM

File: packages/flame_test/CHANGELOG.md (hunk 5 above)
Comment (erickzanardo) [reply packages/flame_test/CHANGELOG.md:4]:
```suggestion
 - Add `flameTest` and `flameWidgetTest`
```

File: packages/flame_test/example/lib/game.dart (hunk 4 above)
Comment (erickzanardo) [reply packages/flame_test/example/lib/game.dart:16]:
```suggestion
    await add(SpriteComponent(sprite: citySprite, size: Vector2.all(200)));
```

**Pull request merged**